src/simulator/halo_simulator.py

```diff
@@ -0,0 +1,217 @@
+
+    def true_reach_by_spend(
+        self, spends: List[float], max_frequency: int = 1
+    ) -> ReachPoint:
+        """Returns the true reach obtained for a given spend vector.
+
+        The true reach is the reach that would have actually been obtained
+        for a given spend vector.  This is in contrast to the simulated
+        reach, which is the noised reach estimated by SMPC.
+
+        Args:
+            spends:  The hypothetical amount spend vector, equal in length to
+              the number of publishers.  spends[i] is the amount that is
+              spent with publisher i.
+            max_frequency:  The maximum frequency for which to report reach.
+        Returns:
+            A ReachPoint representing the true reach that would have been
+            obtained for this spend allocation.
+        """
+        return self._data_set.reach_by_spend(spends, max_frequency)
+
+    def simulated_reach_by_spend(
+        self,
+        spends: List[float],
+        budget: PrivacyBudget,
+        epsilon_split: float = 0.5,
+        max_frequency: int = 1,
+    ) -> ReachPoint:
+        """Returns a simulated differentially private reach estimate.
+
+        Args:
+            spends:  The hypothetical amount spend vector, equal in length to
+              the number of publishers.  spends[i] is the amount that is
+              spent with publisher i.
+            budget:  The amount of privacy budget that can be consumed while
+              satisfying the request.
+            epsilon_split:  Specifies the proportion of the privacy budget that
+              should be allocated to reach estimation.  The remainder is allocated
+              to frequency estimation.
+            max_frequency:  The maximum frequency for which to report reach.
+        Returns:
+            A ReachPoint representing the differentially private estiamte of
+            the reach that would have been obtained for this spend allocation.
+            This estimate is obtained by simulating the construction of
+            Liquid Legions sketches, one per publisher, combining them, and
+            adding differentially private noise to the result.
+        """
+        combined_sketch = self._publishers[0].liquid_legions_sketch(spends[0])
+        estimator = StandardizedHistogramEstimator(
+            max_freq=max_frequency, epsilon=budget.epsilon, epsilon_split=epsilon_split
+        )
+        for i in range(1, len(spends)):
+            sketch = self._publishers[i].liquid_legions_sketch(spends[i])
+            combined_sketch = StandardizedHistogramEstimator.merge_two_sketches(
+                combined_sketch, sketch
+            )
+        frequencies = [
+            round(x) for x in estimator.estimate_cardinality(combined_sketch)
+        ]
+
+        # TODO(jiayu,pasin): Does this look right?
+        self._privacy_tracker.append(
+            NoisingEvent(
+                budget,
+                DP_NOISE_MECHANISM_GAUSSIAN,
```

Comment:
I think this should be either DP_NOISE_MECHANISM_DISCRETE_LAPLACE (as in our paper) or DP_NOISE_MECHANISM_DISCRETE_GAUSSIAN (as in we have been experimenting) as we cannot add continuous noise directly in MPC.
However, if you are thinking of using DP_NOISE_MECHANISM_GAUSSIAN as a proxy for DP_NOISE_MECHANISM_DISCRETE_GAUSSIAN for now, then this looks good.

Reply:
Changed to DP_NOISE_MECHANISM_DISCRETE_GAUSSIAN.